api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java

```diff
@@ -0,0 +1,51 @@
+package net.countercraft.movecraft.craft.datatag;
+
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class CraftDataTagContainer extends HashMap<CraftDataTagKey<?>, Object> {
+
+    public static final Map<NamespacedKey, CraftDataTagKey<?>> REGISTERED_TAGS = new HashMap<>();
+
+    public static <T extends ICraftDataTag> CraftDataTagKey<T> tryRegisterTagKey(final NamespacedKey key, final Supplier<T> supplier) {
+        if (REGISTERED_TAGS.containsKey(key)) {
+            throw new IllegalArgumentException("Duplicate keys are not allowed!");
+        } else {
+            CraftDataTagKey<T> result = new CraftDataTagKey<T>(key, supplier);
+            REGISTERED_TAGS.put(key, result);
+            return result;
+        }
+    }
```

Comment:
Well i do that to avoid stuff like this:
CraftDataTagKey<String> A = REGISTER("bob", ...);
CraftDataTagKey<Boolean> B = REGISTER("bob", ...)

Now, when i want to retrieve A using .get() it will throw some casting exception. ALso the way this works, the registered tags basically act like a default value creation supplier for the map.